File: src/internal/client/agent_groups.go
package client

import (
	"context"
	"fmt"
)

type AgentGroup struct {
	ID           int    `json:"id"`
	Name         string `json:"name"`
	OwnerID      int    `json:"owner_id"`
	Owner        string `json:"owner"`
	Shared       int    `json:"shared"`
	AgentsCount  int    `json:"agents_count"`
	CreationDate int    `json:"creation_date"`
	Timestamp    int    `json:"timestamp"`
}

type AgentGroupCreateRequest struct {
	Name string `json:"name"`
}

type AgentGroupsListResponse struct {
	Groups []AgentGroup `json:"groups"`
}

func (c *Client) CreateAgentGroup(ctx context.Context, name string) (*AgentGroup, error) {
	var resp AgentGroup
	if err := c.Post(ctx, "/scanners/null/agent-groups", AgentGroupCreateRequest{Name: name}, &resp); err != nil {
		return nil, fmt.Errorf("creating agent group: %w", err)
	}
	return &resp, nil
}

func (c *Client) GetAgentGroup(ctx context.Context, groupID int) (*AgentGroup, error) {
	var resp AgentGroup
	if err := c.Get(ctx, fmt.Sprintf("/scanners/null/agent-groups/%d", groupID), &resp); err != nil {
		return nil, fmt.Errorf("getting agent group: %w", err)
	}
	return &resp, nil
}

func (c *Client) DeleteAgentGroup(ctx context.Context, groupID int) error {
	if err := c.Delete(ctx, fmt.Sprintf("/scanners/null/agent-groups/%d", groupID)); err != nil {
		return fmt.Errorf("deleting agent group: %w", err)
	}
	return nil
}

func (c *Client) ListAgentGroups(ctx context.Context) (*AgentGroupsListResponse, error) {
	var resp AgentGroupsListResponse
	if err := c.Get(ctx, "/scanners/null/agent-groups", &resp); err != nil {
		return nil, fmt.Errorf("listing agent groups: %w", err)
	}
	return &resp, nil
}

File: mockapi/tenableio_mock/routers/agent_groups.py
"""Agent groups.

These live under ``/scanners/null/agent-groups``. The literal ``null`` in the
path is Tenable.io's own spelling for "the container's cloud scanner", not a
placeholder the caller is meant to substitute.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Request, Response

from ..errors import bad_request, not_found
from ..models import AgentGroupCreate
from ..store import unix_seconds
from ._common import get_settings, get_store, reject_unknown

router = APIRouter(tags=["agent-groups"])

PREFIX = "/scanners/null/agent-groups"


@router.post(PREFIX)
async def create_agent_group(body: AgentGroupCreate, request: Request) -> dict[str, Any]:
    store, settings = get_store(request), get_settings(request)
    reject_unknown(body, settings)
    if not body.name or not body.name.strip():
        raise bad_request("The agent group name is required.")

    with store.lock:
        if any(g["name"] == body.name for g in store.agent_groups.values()):
            raise bad_request(
                "An agent group with the name you specified already exists.",
                code="duplicate",
            )
        stamp = unix_seconds(store.now())
        group = {
            "id": store.next_id(),
            "name": body.name,
            "owner_id": 1,
            "owner": settings.user,
            "shared": 1,
            "agents_count": 0,
            "creation_date": stamp,
            "timestamp": stamp,
        }
        store.agent_groups[group["id"]] = group
        return dict(group)


@router.get(PREFIX)
async def list_agent_groups(request: Request) -> dict[str, Any]:
    store = get_store(request)
    with store.lock:
        return {"groups": [dict(g) for g in store.agent_groups.values()]}


@router.get(PREFIX + "/{group_id}")
async def get_agent_group(group_id: int, request: Request) -> dict[str, Any]:
    store = get_store(request)
    with store.lock:
        group = store.agent_groups.get(group_id)
        if group is None:
            raise not_found(f"Agent group {group_id} was not found.")
        return dict(group)


@router.delete(PREFIX + "/{group_id}")
async def delete_agent_group(group_id: int, request: Request) -> Response:
    store = get_store(request)
    with store.lock:
        if group_id not in store.agent_groups:
            raise not_found(f"Agent group {group_id} was not found.")
        del store.agent_groups[group_id]
    return Response(status_code=200)
